pydra/engine/core.py

```diff
@@ -281,13 +281,15 @@ def checksum_states(self, state_index=None):
 
         """
         if is_workflow(self) and self.inputs._graph_checksums is attr.NOTHING:
-            self.inputs._graph_checksums = [nd.checksum for nd in self.graph_sorted]
+            self.inputs._graph_checksums = {
+                nd.name: nd.checksum for nd in self.graph_sorted
+            }
 
         if state_index is not None:
-            inputs_copy = deepcopy(self.inputs)
+            inputs_copy = copy(self.inputs)
```

Comment:
are we sure, that removing deepcopy here will be always ok?

Reply:
I'm not 100% but pretty confident. It looks like the copy is only done so that the state-array values can be resolved to single values on L294

```
        if state_index is not None:
            inputs_copy = copy(self.inputs)
            for key, ind in self.state.inputs_ind[state_index].items():
                val = self._extract_input_el(
                    inputs=self.inputs, inp_nm=key.split(".")[1], ind=ind
                )
                setattr(inputs_copy, key.split(".")[1], val)
```

Reply:
https://github.com/nipype/pydra/blob/4e1d4a89f40857453ac8cb86f1061426a8ebcb3e/pydra/engine/core.py#L288-L294

Reply:
yes, I think that was indeed mostly needed for the tasks with states